backend/src/main/scala/bloop/Compiler.scala

```diff
@@ -357,13 +366,74 @@ object Compiler {
 
     val uniqueInputs = compileInputs.uniqueInputs
     reporter.reportStartCompilation(previousProblems)
+    val fileManager = newFileManager
+
+    // Manually skip redundant best-effort compilations. This is necessary because compiler
+    // phases supplying the data needed to skip compilations in zinc remain unimplemented for now.
+    val noopBestEffortResult = compileInputs.previousCompilerResult match {
+      case Failed(
+            problems,
+            t,
+            elapsed,
+            _,
+            bestEffortProducts @ Some(BestEffortProducts(compileProducts, previousHash))
+          ) if isBestEffortMode =>
+        val newHash = BestEffortUtils.hashResult(
+          compileProducts.newClassesDir,
+          compileInputs.sources,
+          compileInputs.classpath
+        )
+
+        if (newHash == previousHash) {
+          reporter.processEndCompilation(
+            problems,
+            bsp.StatusCode.Error,
+            None,
+            None
+          )
+          reporter.reportEndCompilation()
+          val backgroundTasks = new CompileBackgroundTasks {
+            def trigger(
+                clientClassesObserver: ClientClassesObserver,
+                clientReporter: Reporter,
+                clientTracer: BraveTracer,
+                clientLogger: Logger
+            ): Task[Unit] = Task.defer {
+              val clientClassesDir = clientClassesObserver.classesDir
+              clientLogger.debug(s"Triggering background tasks for $clientClassesDir")
+              val firstTask = Task { BloopPaths.delete(AbsolutePath(newClassesDir)) }
+              val secondTask = updateExternalClassesDirWithReadOnly(
+                clientClassesDir,
+                clientTracer,
+                clientLogger,
+                compileInputs,
+                compileProducts.newClassesDir,
+                readOnlyCopyDenylist = mutable.HashSet.empty,
+                allInvalidatedClassFilesForProject,
+                allInvalidatedExtraCompileProducts
+              )
+              Task
+                .gatherUnordered(List(firstTask, secondTask))
+                .map(_ => ())
+            }
+          }
+          Some(Failed(problems, t, elapsed, backgroundTasks, bestEffortProducts))
+        } else None
+      case _ => None
+    }
+
+    if (noopBestEffortResult.isDefined) {
+      logger.debug("Skipping redundant best-effort compilation")
+      return Task { noopBestEffortResult.get }
```

Comment:
This is probably a bit ugly as nonLocal returns are not used anywhere else in bloop, but any other solution would destroy the diff in Compiler.scala. Maybe I should squash the commits into one and then add a change from this to an else/pattern match, so that the review can be done on the basis of the first commit, but the actual merged code will be cleaner?

Reply:
Might be good to squash as it's already a lot of commits with contradicting changes